runtime/src/main/java/org/corfudb/util/concurrent/SingletonResource.java

```diff
@@ -1,102 +1,78 @@
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
+
 import javax.annotation.Nonnull;
 
-/** Utility class which implements a singleton resource pattern.
+/**
+ * Utility class which implements a singleton resource pattern.
  *
- *  <p>A {@link SingletonResource} is a common resource pattern where the first thread that
- *  needs to use a resource instantiates it. Subsequent threads should re-use the resource
- *  instantiated by the first thread. The {@link SingletonResource#cleanup(Consumer)} allows
- *  the developer to release resources if the resource has been instantiated (doing nothing
- *  if no thread ever created the resource in the first place).
+ * <p>A {@link SingletonResource} is a common resource pattern where the first thread that
+ * needs to use a resource instantiates it. Subsequent threads should re-use the resource
+ * instantiated by the first thread. The {@link SingletonResource#cleanup(Consumer)} allows
+ * the developer to release resources if the resource has been instantiated (doing nothing
+ * if no thread ever created the resource in the first place).
  *
- * @param <T>   The type of resource this {@link SingletonResource} holds.
+ * @param <T> The type of resource this {@link SingletonResource} holds.
  */
 public class SingletonResource<T> {
 
-    /** The resource to be held. */
-    private T resource;
-
-    /** A generator which provides the resource. */
+    /**
+     * A generator which provides the resource.
+     */
     private final Supplier<T> generator;
 
-    /** A stamped lock which controls access to the resource. */
-    private final StampedLock lock;
+    /**
+     * The resource to be held.
+     */
+    private final AtomicReference<T> resource;
 
-    /** Factory method with similar semantics as a {@link ThreadLocal}.
+    /**
+     * Factory method with similar semantics as a {@link ThreadLocal}.
      *
-     * @param generator     A method to be called when a new {@link R} is needed.
-     * @param <R>           The type of the resource to be provided.
-     * @return              A new {@link SingletonResource}.
+     * @param generator A method to be called when a new {@link R} is needed.
+     * @param <R>       The type of the resource to be provided.
+     * @return A new {@link SingletonResource}.
      */
     public static <R> SingletonResource<R> withInitial(@Nonnull Supplier<R> generator) {
         return new SingletonResource<>(generator);
     }
 
-    /** Generate a new {@link SingletonResource}.
+    /**
+     * Generate a new {@link SingletonResource}.
      *
-     * @param generator     A method to be called when a new {@link T} is needed.
+     * @param generator A method to be called when a new {@link T} is needed.
      */
     private SingletonResource(Supplier<T> generator) {
-        lock = new StampedLock();
+        this.resource = new AtomicReference<T>();
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace the type specification in this constructor call with the diamond operator ("<>"). [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2293)